### ISSUE TITLE
rustdoc: nested items should not be documented with `--document-private-items`

### ISSUE THREAD

**Issue by GuillaumeGomez** [rust-lang/rust#110422]
If we run `rustdoc` with the `--document-private-items` option on this code:

```rust
pub fn foo() {
    fn bar() {}
}
```

Both `foo` and `bar` are documented, but `bar` not being accessible outside of `foo`, it shouldn't be documented.

**Comment by compiler-errors** [comment]
@GuillaumeGomez probably knows this, but for others seeing this issue, I think it's probably due to #104889 (which was reverted in #107083, and later re-applied in https://github.com/rust-lang/rust/pull/107000#issuecomment-1405407485).

It has to do with the fact that we're using a HIR visitor now, so we're visiting nested items unconditionally.

**Comment by GuillaumeGomez** [comment]
Yes I know where it's coming from. :smile: 

I was just thinking today that since we were going through the whole HIR without checking whether or not we were in a block, all elements inside the block could be visible. Tested and... :fearful: 

Should be pretty easy to fix.

**Comment by compiler-errors** [comment]
For reference, I put up https://github.com/serde-rs/serde/pull/2426 because this was causing documentation to become really verbose with serde + `--document-private-items`. If you fix this such that we no longer document those items, we could probably revert that PR.

**Comment by GuillaumeGomez** [comment]
Indeed. Don't hesitate to come to me if you have such issues with rustdoc. Such problems should definitely not go unnoticed.

**Comment by compiler-errors** [comment]
> Don't hesitate to come to me if you have such issues with rustdoc. 

The thing is, it was not particularly clear from the PRs whether this was intentional or not.

**Comment by GuillaumeGomez** [comment]
Unfortunately it wasn't. The goal was to retrieve the missing impl blocks. But if anything isn't clear, don't hesitate to ask too. :wink:
